### ISSUE TITLE
SumReducer doesn't respect user-defined operator+ for adding compatible types

### ISSUE THREAD

**Issue by wrwilliams** [LLNL/RAJA#357]
This code:

```
    struct FrameSet : public std::set<ParseFrame*>
    {
        FrameSet(unsigned int = 0) {}
        template <typename ContainerT>
        FrameSet operator+(const ContainerT& c) const
        {
         FrameSet result(*this);
         result.insert(c.begin(), c.end());
         return result;
        }
    };
```
should allow a FrameSet to accept any container of ParseFrame* as the right hand side of a reduction. However, `operator+=` requires its argument to be of the same type (or convertible to the same type) as the reduction value. Adding an `operator+=` overload to the type being reduced is also no help, as SumReducer is only implemented in terms of `operator+`.

I'd definitely make the template parameter for operator+= independent, allowing anything T considers valid for addition to be reduced. I'd at least consider an implementation based on T::operator+= as an overload.

Tagging @trws @willkill07 as requested in #356 .

**Comment by willkill07** [comment]
@wrwilliams thanks for the tag. Unless FrameSet has `operator+(FrameSet)` defined, reductions will NOT work. A value must be trivially constructible and support whatever the reduction operator is (be it +, overload of min, or overload of max).

It is possible to add the template argument for the reduction implementation and just perfect forward it through the machinery. I’ll look into this. But due to thread-local copies of FrameSet needing combined, the aforementioned self-addition needs to also be well defined.

**Comment by willkill07** [comment]
We also have this concept of an identity value. I just need to improve the reduction documentation 😅

**Comment by wrwilliams** [comment]
> We also have this concept of an identity value. I just need to improve the reduction documentation :sweat_smile:

Yeah, as far as I could tell, an identity for SumReduce is defined in terms of "what's constructible from 0", with an implicit assumption that all sums are equivalent to numeric math (thus the constructor taking an int).

Note that with this template defined, as long as I ensure we're adding two FrameSets, things work. I didn't expect reductions to solely accept T + U without T + T also being defined; I just expected overloaded addition to work as well. Conversion operators/implicit constructors should already be legal as the code stands but they're not going to be as efficient as an addition overload for nontrivial types.
